server/builtin/app.go

```diff
@@ -0,0 +1,273 @@
+				return
+			}
+			err = nil
+			out = ioutil.NopCloser(bytes.NewReader(data))
+		}
+	}(a.conf.Logger())
+
+	f, ok := a.router[creq.Path]
+	if !ok {
+		return nil, utils.NewNotFoundError("%s is not found", creq.Path)
+	}
+	cresp := f(creq)
+	data, err := json.Marshal(cresp)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.NopCloser(bytes.NewReader(data)), nil
+}
+
+func (a *builtinApp) GetStatic(_ apps.App, path string) (io.ReadCloser, int, error) {
+	return nil, http.StatusNotFound, utils.NewNotFoundError("static support is not implemented")
+}
+
+func mdResponse(format string, args ...interface{}) apps.CallResponse {
+	return apps.CallResponse{
+		Type:     apps.CallResponseTypeOK,
+		Markdown: fmt.Sprintf(format, args...),
+	}
+}
+
+func formResponse(form apps.Form) apps.CallResponse {
+	return apps.CallResponse{
+		Type: apps.CallResponseTypeForm,
+		Form: &form,
+	}
+}
+
+func dataResponse(data interface{}) apps.CallResponse {
+	return apps.CallResponse{
+		Type: apps.CallResponseTypeOK,
+		Data: data,
+	}
+}
+
+func emptyForm(_ apps.CallRequest) apps.CallResponse {
+	return formResponse(apps.Form{})
+}
+
+func submitPath(p string) string {
+	return path.Join(p, "submit")
+}
+
+func formPath(p string) string {
+	return path.Join(p, "form")
+}
+
+func lookupPath(p string) string {
+	return path.Join(p, "lookup")
+}
+
+const SysadminOnly = true
+
+func (a *builtinApp) handle(
+	path string,
+	sysadminOnly bool,
+	submitf func(apps.CallRequest) apps.CallResponse,
+) {
+	a.router[submitPath(path)] = requireSysadmin(sysadminOnly, submitf)
+}
+
+func (a *builtinApp) withLookup(
+	path string,
+	sysadminOnly bool,
+	submitf func(apps.CallRequest) apps.CallResponse,
+	lookupf func(apps.CallRequest) ([]apps.SelectOption, error),
+) {
+	a.handle(path, sysadminOnly, submitf)
+
+	if lookupf == nil {
+		return
+	}
+	type lookupResponse struct {
+		Items []apps.SelectOption `json:"items"`
+	}
+	a.router[lookupPath(path)] = requireSysadmin(sysadminOnly,
+		func(creq apps.CallRequest) apps.CallResponse {
+			opts, err := lookupf(creq)
+			if err != nil {
+				return apps.NewErrorCallResponse(err)
+			}
+			return dataResponse(lookupResponse{opts})
+		})
+}
+
+func (a *builtinApp) withForm(
+	path string,
+	sysadminOnly bool,
+	submitf func(apps.CallRequest) apps.CallResponse,
+	formf func(apps.CallRequest) (*apps.Form, error),
+) {
+	a.handle(path, sysadminOnly, submitf)
+
+	if formf == nil {
+		return
+	}
+	a.router[lookupPath(path)] = requireSysadmin(sysadminOnly,
+		func(creq apps.CallRequest) apps.CallResponse {
+			form, err := formf(creq)
+			if err != nil {
+				return apps.NewErrorCallResponse(err)
+			}
+			return formResponse(*form)
+		})
+}
+
+func requireSysadmin(require bool, handler func(apps.CallRequest) apps.CallResponse) func(apps.CallRequest) apps.CallResponse {
+	if !require {
+		return handler
+	}
+	return func(creq apps.CallRequest) apps.CallResponse {
+		if creq.Context.ExpandedContext.AdminAccessToken == "" {
```

Comment:
0/5. Do you feel it is better to expand the token instead of expanding the user and checking its role? The token seems more "dangerous" than the user object.

Reply:
0/5 the token's presence is a single bit, it can be explicitly expressed with the call definition... expanding the user is "fatter" (more data), and a little opaque (all Calls are passed to the builtin, rather than having the unauthorized rejected by the proxy itself, based on the `Expand` declaration... And we'd still need an `Expand`... so it's not more compact.